lib/list_versions.go

```diff
@@ -1,142 +1,166 @@
-			if r.MatchString(versions[i]) {
-				str := r.FindString(versions[i])
-				trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-				return trimstr, nil
-			}
-		}
-	} else if preRelease == false {
-		listAll := false
-		tflist, _ := GetTFList(mirrorURL, listAll) //get list of versions
-		version = fmt.Sprintf("~> %v", version)
-		semv, err := SemVerParser(&version, tflist)
+//GetTFReleases :  Get all terraform releases given the hashicorp url
+func GetTFReleases(mirrorURL string, preRelease bool) ([]*Release, error) {
+	limit := 20
+	queryParams := map[string]string{"limit": strconv.Itoa(limit)}
+	releaseSet, err := getReleases(mirrorURL, queryParams)
+	if err != nil {
+		return nil, err
+	}
+	var releases []*Release
+	releases = append(releases, releaseSet...)
+	for len(releaseSet) == limit {
+		queryParams["after"] = releaseSet[len(releaseSet)-1].TimestampCreated.String()
+		releaseSet, err = getReleases(mirrorURL, queryParams)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		return semv, nil
+		releases = append(releases, releaseSet...)
 	}
-	return "", nil
-}
-
-//GetTFURLBody : Get list of terraform versions from hashicorp releases
-func GetTFURLBody(mirrorURL string) ([]string, error) {
 
-	hasSlash := strings.HasSuffix(mirrorURL, "/")
-	if !hasSlash { //if does not have slash - append slash
-		mirrorURL = fmt.Sprintf("%s/", mirrorURL)
+	if !preRelease {
+		releases = removePreReleases(releases)
 	}
-	resp, errURL := http.Get(mirrorURL)
+	sort.Slice(releases, func(i, j int) bool {
+		return releases[i].Version > releases[j].Version
+	})
+	return releases, nil
+}
+
+//GetTFRelease :  Get the requested terraform release given the hashicorp url
+func GetTFRelease(mirrorURL, requestedVersion string) (*Release, error) {
+	resp, errURL := httpGet(mirrorURL+"/"+requestedVersion, nil)
 	if errURL != nil {
-		log.Printf("[Error] : Getting url: %v", errURL)
-		os.Exit(1)
-		return nil, errURL
+		return nil, fmt.Errorf("[Error] : Getting url: %v", errURL)
 	}
 	defer resp.Body.Close()
 
-	if resp.StatusCode != 200 {
-		log.Printf("[Error] : Retrieving contents from url: %s", mirrorURL)
-		os.Exit(1)
+	body := new(bytes.Buffer)
+	if _, err := io.Copy(body, resp.Body); err != nil {
+		return nil, err
 	}
-
-	body, errBody := ioutil.ReadAll(resp.Body)
-	if errBody != nil {
-		log.Printf("[Error] : reading body: %v", errBody)
-		os.Exit(1)
-		return nil, errBody
+	var release *Release
+	if err := json.Unmarshal(body.Bytes(), &release); err != nil {
+		return nil, fmt.Errorf("%s: %s", err, body.String())
 	}
+	return release, nil
 
-	bodyString := string(body)
-	result := strings.Split(bodyString, "\n")
+}
 
-	return result, nil
+//removePreReleases : Removes any prerelease versions from a given slice of Release.
+func removePreReleases(releases []*Release) []*Release {
+	for i, r := range releases {
+		if r.IsPrerelease {
+			releases = append(releases[:i], releases[i+1:]...)
+		}
+	}
+	return releases
```

Comment:
I don't think this is working as you expect it to, this pattern can work if you do it in reverse, but not iterating forward:
https://go.dev/play/p/P9OiqbT-HRp



A simpler pattern is to just build up a new slice as you go
```suggestion
//removePreReleases : Removes any prerelease versions from a given slice of Release.
func removePreReleases(releases []*Release) []*Release {
	realReleases := []*Release{}
	for i, r := range releases {
		if !r.IsPrerelease {
			realReleases = append(realReleases, releases[i])
		}
	}
	return realReleases
```

Reply:
Ah nice, thanks!